### PR TITLE
Clean up cockpitunixfd.[hc], move to Linux close_range API, various simplifications

### PR DESCRIPTION
**Pull request by martinpitt**
Linux 5.9 finally added a close_range() call [1]. This is not yet
exposed by glibc, but let's move our code to be ready to use it once it
is available. Add a new common cockpit_close_range() function which
offers the same API, and use it everywhere.

Remove cockpit_unix_fd_close_*() from cockpitunixfd.h. It does not fit
there, as (1) that file otherwise contains backports of [2] and will
eventually disappear, and (2) we also want to use close_range() from
programs that don't include glib.

This also gets rid of some fdwalk() implementation duplication.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=278a5fbaed89dacd04e9d
[2] https://developer.gnome.org/glib/stable/glib-UNIX-specific-utilities-and-integration.html

### CODE REVIEW

**Comment by martinpitt** [comment]
As Lis wants to postpone #16156, I rebased this to current main. This should not interfere any more with anything else now.

**Comment by martinpitt** [comment]
I think it's worth landing this now:

 * We remove the triplication of code (I diffed the implementations to make sure they do the same thing)
 * We avoid introducing more code that uses the old API (in particular with the `except` thing, which `close_range()` does not support)
 * This gives us a central place to introduce `close_range()`, with the `#ifdef HAVE_CLOSE_RANGE` (for old glibcs) and handling `ENOSYS` (for old kernels). We'll need these fallbacks for the next few years for at least RHEL/CentOS 8 and older Debians/Ubuntus, but I don't want to wait that long for using `close_range()` once it becomes available.

File: src/bridge/bridge.c
Comment (allisonkarlitskaya) [inline src/bridge/bridge.c:177]: This is begging for a rewrite: hardcode `--print-address=3` and `dup2()` the fd over, and just close everything from 4 and up.

Even better: use `GSubproces` instead, with `g_subprocess_launcher_take_fd()`, and drop the child setup function.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:107]: Just drop the flags parameter, if you're going to do that.  This is an internal API after all.

Drop the `max_fd` paramter while you're at it, since there is one user of it that should be refactored anyway.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:38]: I have no idea on what kind of Linux system we're supposed to have `fdwalk()`, but it's not a part of glibc.  Please drop the conditionalization of this function and the corresponding check for it in `configure.ac`.

Meanwhile, the callback structure here is actually pointless and has always bothered me: merge the struct typedef plus all three functions in this file into a single simple function, with a single parameter, which does one thing: closes all fds above a certain number.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
I like the -339, but I think we can do better than +137 :)

File: src/pam-ssh-add/Makefile.am
Comment (allisonkarlitskaya) [inline src/pam-ssh-add/Makefile.am:16]: Well, that's nice I guess?

File: src/pam-ssh-add/Makefile.am
Comment (allisonkarlitskaya) [inline src/pam-ssh-add/Makefile.am:10]: ...and what's this?

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
These either slipped in by accident, or belong to a separate commit.  Either way, more explanation please :)

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:107]: Nack -- the *point* was to move the `close_range()` API, which has this very API. As the next step this will call the actual `close_range()` once it's available in glibc, and keep the current code as a fallback, and eventually we can entirely drop it.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:38]: I can add simplification of fdwalk to my TODO list, but I don't want to do it in this PR -- this cleanly moves code around withouth changing behaviour or implementation.

File: src/pam-ssh-add/Makefile.am
Comment (martinpitt) [reply src/pam-ssh-add/Makefile.am:16]: Right, this is unrelated cleanup. Split out into a separate commit.

File: src/pam-ssh-add/Makefile.am
Comment (martinpitt) [reply src/pam-ssh-add/Makefile.am:10]: It was a bug in the Makefile which didn't surface until trying to add a second .c file. Fixed in a separate commit.

**Comment by martinpitt** [comment]
> I like the -339, but I think we can do better than +137 :)

Have a look at PR #16164 which is queued after this one -- it removes even more :grin:

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:38]: I did add a commit which drops the configure check, this is easy enough and safe.

Comment (martinpitt) [reply src/common/cockpitcloserange.c:38]: OK, on second thought I also added a commit right here which simplifies the function as you mentioned.

File: src/bridge/bridge.c
Comment (martinpitt) [reply src/bridge/bridge.c:177]: Done. Tested with:
```
❱❱❱ make && env -u DBUS_SESSION_BUS_ADDRESS G_MESSAGES_DEBUG=all ./cockpit-bridge | cat
[...]
(cockpit-bridge:232955): cockpit-bridge-DEBUG: 08:49:54.609: launched dbus-daemon: 232957
(cockpit-bridge:232955): cockpit-bridge-DEBUG: 08:49:54.613: session bus address: unix:abstract=/tmp/dbus-uolitsRRHL,guid=1fe30868a2766df409b2654961024f92
```
... and I can see the process properly in `ps`. This results in 31 fewer lines of code, and looks quite nice indeed. Thanks for pointing out, I didn't know that API yet.

File: src/bridge/bridge.c
Comment (allisonkarlitskaya) [inline src/bridge/bridge.c:196]: Why 4?

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:69]: Just put the condition here and call `close()` directly?

All of the error checking above is a bit weird imho.  `close()` is always successful. Any error you get there is delayed from a previously failed operation (and the fd is still closed).  The obvious exception is EBADF, but even in that case you end up with a closed fd.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:87]: And here, just start the loop from the minimum fd and then you don't need any extra logic.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:41]: May as well assert that the upper range is -1 as well.  It lets us simplify logic some.

File: src/common/cockpitpipe.c
Comment (allisonkarlitskaya) [inline src/common/cockpitpipe.c:1411]: It's a bit weird to see the glib macro used with a "pure C" call.  Maybe that's because you'd need another header for the libc one?

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:1]: Copyright header please, and #include "config.h" as the first line.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:43]: Cockpit is always on Linux....

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:53]: That seems redundant with the error check after the int parsing.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

**Comment by martinpitt** [comment]
Fixed a [memory leak](https://github.com/cockpit-project/cockpit/pull/16157/checks?check_run_id=3189233294)

File: src/bridge/bridge.c
Comment (martinpitt) [reply src/bridge/bridge.c:196]: Right, should have taken 42! (I looked at the number in the original code, and it was 4, that must have stuck in  my brain). Changed to 3.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:43]: Dropped.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:41]: Done.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:69]: Done, thanks for pointing out that close() is always successful in Linux -- that helps indeed!

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:1]: Done, and also added the missing `#pragma once`.

File: src/common/cockpitpipe.c
Comment (martinpitt) [reply src/common/cockpitpipe.c:1411]: Replaced with `INT_MAX`.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c]: I validated that this works fine, strtol()'s end pointer is indeed at the `.` then (it's not trying funny float parsing stuff). Dropped.

**Comment by martinpitt** [comment]
@allisonkarlitskaya : Addressed all your comments, thanks for your great review!

File: src/bridge/bridge.c
Comment (martinpitt) [inline src/bridge/bridge.c:440]: FTR: once this gets a +1 and lands, I'm happy to rewrite the SSH agent setup as well, as right now it's asymmetric. But I wanted to get one right first.

File: src/bridge/bridge.c
Comment (allisonkarlitskaya) [inline src/bridge/bridge.c:183]: if you're gonna be juggling this all around anyway, can you move it to the style of declaration the variable at its site of first use?

File: src/bridge/bridge.c
Comment (allisonkarlitskaya) [inline src/bridge/bridge.c:192]: also stdout?  I wonder if we couldn't use a stdout pipe: that would be a lot easier than all of this messing about with launchers and pipes and read loops...

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:36]: move variables down to where we use them.  in the case of `fd` you'll have it twice, but that's good.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:73]: You always return 0 now.  Make this void and drop the error handling from the callers.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: src/bridge/bridge.c
Comment (martinpitt) [reply src/bridge/bridge.c:183]: Done.

File: src/bridge/bridge.c
Comment (martinpitt) [reply src/bridge/bridge.c:181]: The current `dbus-daemon` that I have works ok with `--print-address` (which goes to stdout). I'm a bit afraid of dbus-daemon writing anything else there, but given that it is advertised for exactly this use case, let's do this. I do this as a separate commit, just in case.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c]: Done.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:68]: close_range() returns an int, I'd like to keep that API.

**Comment by martinpitt** [comment]
I already retried packit builds once, it seems [centos-8-stream COPR builds](https://download.copr.fedorainfracloud.org/results/packit/cockpit-project-cockpit-16157/centos-stream-8-x86_64/02352532-cockpit/builder-live.log.gz) have some ongoing networking problems. This is unrelated though, and our own [centos-8-stream](https://logs.cockpit-project.org/logs/pull-16157-20210729-085640-9a2f130b-centos-8-stream/log.html) build shows that there are no API incompatibilities.

File: src/bridge/bridge.c
Comment (allisonkarlitskaya) [inline src/bridge/bridge.c:203]: `read_all()`

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: src/bridge/bridge.c
Comment (martinpitt) [reply src/bridge/bridge.c:203]: Can't -- dbus-daemon does not close its stdout after writing the address, so read_all() will hang forever (I did try this). I'm afraid we have to read the full line. I was a bit disappointed that there's no `g_input_stream_read_line()` as an fgets() counterpart, that'd be really handy.

Comment (allisonkarlitskaya) [reply src/bridge/bridge.c:203]: `g_data_input_stream_read_line ()`

Comment (martinpitt) [reply src/bridge/bridge.c]: Hah, thanks! So much nicer now. Pushed.

**Comment by martinpitt** [comment]
This is just a bit short of `g_spawn_subprocess_and_read_one_line_from_it()` :grin:

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Code looks good, but after all of that back and forth this needs a bit of rebasing: let's do the GSubprocess cleanups first and then do the implementation of the close_range() call in a single commit, in its current form.  There's no sense doing the more complicated implementation only to simplify it a couple of commits later.

Also: one of your commit messages talks about "stout" instead of "stdout".  Time for a bear? :)

**Comment by martinpitt** [comment]
@allisonkarlitskaya : Rebased in the suggested order, and validated that it was a net-zero change. As this is such a grab-bag of cleanups now, I also added the removal of cockpitunixfd.[hc], as that's fairly simple now.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:21]: Usually a newline between `config.h` and the other includes.

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [inline src/common/cockpitcloserange.c:50]:
```suggestion
          long l = strtol (de->d_name, &e, 10);
```

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
trivial

File: src/common/cockpitcloserange.c
Comment (allisonkarlitskaya) [reply src/common/cockpitcloserange.c:50]: also: pretty sure you don't need to initialize `e`.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c:22]: Done.

File: src/common/cockpitcloserange.c
Comment (martinpitt) [reply src/common/cockpitcloserange.c]: Done both.

**Review by allisonkarlitskaya — APPROVED** [review]
tja.

^5.

**Pull request merged**
